### PR TITLE
Integration with magento - first part

### PR DESCRIPTION
**Pull request by TomaszQr**

### CODE REVIEW

File: app/Integrations/Magento/Connector/Token.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	 * @var string
+	 */
+	private $token;
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function authorize()
+	{
+		$response = (new \GuzzleHttp\Client())->post(\App\Config::component('Magento', 'ADDRESS_API') . '/rest/V1/integration/admin/token', [
```
Comment (rskrzypczak) [inline app/Integrations/Magento/Connector/Token.php]: \App\RequestHttp::getOptions()

File: config/Components/ConfigTemplates.php [hunk 2]
```diff
@@ -222,5 +222,27 @@
 			'default' => [],
 			'description' => 'List of modules for which Twitter has been enabled.',
 		]
-	]
+	],
+	'Magento' => [
+		'CONNECTOR' => [
```
Comment (rskrzypczak) [inline config/Components/ConfigTemplates.php]: camelCase

File: app/Integrations/Magento/Connector/Token.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
+<?php
+/**
+ * Connector based on session.
+ *
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Connector/Token.php]: @Package Integration

File: app/Integrations/Magento/Connector/Token.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	 * {@inheritdoc}
+	 */
+	public function authorize()
+	{
+		$response = (new \GuzzleHttp\Client())->post(\App\Config::component('Magento', 'ADDRESS_API') . '/rest/V1/integration/admin/token', \App\RequestHttp::getOptions() + [
+			'json' => ['username' => \App\Config::component('Magento', 'USERNAME'), 'password' => \App\Config::component('Magento', 'PASSWORD')]]);
+		if ($response->getStatusCode() !== 200) {
+			throw new AppException();
+		}
+		$this->token =  str_replace('"', '', (string) $response->getBody());
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Connector/Token.php]: json_decode  ?

File: app/Integrations/Magento/Connector/Token.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function request(string $method, string $action, array $params = []): string
+	{
+		$response = (new \GuzzleHttp\Client())->request($method, \App\Config::component('Magento', 'ADDRESS_API') . $action, \App\RequestHttp::getOptions() + [
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Connector/Token.php]:
```suggestion
		$response = (new \GuzzleHttp\Client(\App\RequestHttp::getOptions()))->request($method, \App\Config::component('Magento', 'ADDRESS_API') . $action,[
```

File: app/Integrations/Magento/Connector/Token.php [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+			'headers' => [
+				'user-agent' => 'YetiForceCRM/' . \App\Version::get(),
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Connector/Token.php]:
```suggestion
```

File: app/Integrations/Magento/Connector/Token.php [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+	 * @var string
+	 */
+	private $token;
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function authorize()
+	{
+		$response = (new \GuzzleHttp\Client())->post(\App\Config::component('Magento', 'ADDRESS_API') . '/rest/V1/integration/admin/token', \App\RequestHttp::getOptions() + [
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Connector/Token.php]: GuzzleHttp\Client should be in try / catch

File: app/Integrations/Magento/Connector/Token.php (hunk 5 above)
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Connector/Token.php]: GuzzleHttp\Client should be in try / catch

File: app/Integrations/Magento/Magento.php [hunk 8]
```diff
@@ -0,0 +1,63 @@
+<?php
+/**
+ * Main class to integration with magento.
+ *
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Magento.php]: @Package Integration

File: app/Integrations/Magento/Magento.php [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+namespace App\Integrations\Magento;
+
+use App\Exceptions\AppException;
+use App\Integrations\Magento\Connector\ConnectorInterface;
+use App\Integrations\Magento\Synchronizator\Category;
+
+/**
+ * Magento class.
+ */
+class Magento
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Magento.php]:
```suggestion
class Controller
```

File: app/Integrations/Magento/Magento.php [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+	/**
+	 * Returns connector.
+	 *
+	 * @return object
+	 */
+	public function getConnector()
+	{
+		if (empty($this->connector)) {
+			$className = '\\App\\Integrations\\Magento\\Connector\\' . \App\Config::component('Magento', 'CONNECTOR', 'Token');
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Magento.php]: add \class_exists

File: app/Integrations/Magento/Magento.php [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+		return $this->connector;
+	}
+
+	/**
+	 * Constructor. Connect with magento and authorize.
+	 */
+	public function __construct()
+	{
+		$connector = $this->getConnector();
+		$connector->authorize();
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Magento.php]:
```suggestion
		$this->getConnector()->authorize();
```

File: app/Integrations/Magento/Synchronizator/Base.php [hunk 12]
```diff
@@ -0,0 +1,42 @@
+<?php
+/**
+ * Synchronize.
+ *
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Synchronizator/Base.php]: @Package Integration

File: app/Integrations/Magento/Magento.php [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+use App\Integrations\Magento\Connector\ConnectorInterface;
+use App\Integrations\Magento\Synchronizator\Category;
+
+/**
+ * Magento class.
+ */
+class Magento
+{
+	/**
+	 * Connector with magento.
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Magento.php]: type

File: app/Integrations/Magento/Synchronizator/Base.php [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+	 * Connector.
+	 *
+	 * @var object
+	 */
+	protected $connector;
+
+	/**
+	 * Sets connector to communicate with system.
+	 *
+	 * @param object $connector
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Synchronizator/Base.php]: type

File: app/Integrations/Magento/Synchronizator/Category.php [hunk 15]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+/**
+ * Synchronize categories for products.
+ *
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Synchronizator/Category.php]: @Package Integration

File: install/install_schema/scheme.sql [hunk 16]
```diff
@@ -992,6 +992,21 @@ CREATE TABLE `s_yf_companies` (
   PRIMARY KEY (`id`)
 ) ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=utf8;
 
+/*Table structure for table `s_yf_magento_config` */
+
+CREATE TABLE `s_yf_magento_config` (
```
Comment (mariuszkrzaczkowski) [inline install/install_schema/scheme.sql]:
```suggestion
CREATE TABLE `i_yf_magento_config` (
```

File: install/install_schema/scheme.sql [hunk 17]
```diff
@@ -992,6 +992,21 @@ CREATE TABLE `s_yf_companies` (
...
+  `name` varchar(15) DEFAULT NULL,
+  `value` varchar(50) DEFAULT NULL
+) ENGINE=InnoDB DEFAULT CHARSET=utf8;
+
+/*Table structure for table `s_yf_magento_record` */
+
+CREATE TABLE `s_yf_magento_record` (
```
Comment (mariuszkrzaczkowski) [inline install/install_schema/scheme.sql]:
```suggestion
CREATE TABLE `i_yf_magento_record` (
```

File: install/install_schema/scheme.sql [hunk 18]
```diff
@@ -992,6 +992,21 @@ CREATE TABLE `s_yf_companies` (
...
+  `id` int(10) unsigned NOT NULL,
+  `crmid` int(10) unsigned NOT NULL,
```
Comment (mariuszkrzaczkowski) [inline install/install_schema/scheme.sql]: FK to vtiger_crmentity

**Pull request merged**